### PR TITLE
New Sphinx tutorial, part II

### PR DESCRIPTION
**Pull request by astrojuanlu**
### Feature or Bugfix
- Feature

### Purpose
This is a continuation of the tutorial we started at #9276. It expands some sections already started in the first part and adds some new ones, as originally proposed in #9165:

4. First steps to document our project using Sphinx
5. Customizing Sphinx configuration
6. Writing narrative documentation with Sphinx

Again, the idea is to keep these pull requests manageable, iterate in small chunks, and in the meanwhile keep the tutorial in a complete state.

Rendered version: https://sphinx--9355.org.readthedocs.build/en/9355/tutorial/index.html

cc @ericholscher 

### Relates
- #9276
- #9165

### CODE REVIEW

File: doc/tutorial/index.rst [hunk 1]
```diff
@@ -184,6 +195,243 @@ as before, or leverage the convenience script as follows:
...
+
+And then, locate the ``html_theme`` variable on your ``conf.py`` and replace
+its value as follows:
+
+.. code-block:: python
+
+   # The theme to use for HTML and HTML Help pages.  See the documentation for
+   # a list of builtin themes.
+   #
+   html_theme = 'alabaster'
```
Comment (davidfischer) [inline doc/tutorial/index.rst]: Should this be `furo`?

Comment (astrojuanlu) [reply doc/tutorial/index.rst]: Woops good catch!

File: doc/tutorial/index.rst [hunk 2]
```diff
@@ -184,6 +195,243 @@ as before, or leverage the convenience script as follows:
...
+Exporting your documentation to other formats
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Sphinx supports a variety of formats apart from HTML, including PDF, EPUB,
+:ref:`and more <builders>`.  For example, to build your documentation as an
+e-book in EPUB format, run this command from the ``docs`` directory:
+
+.. code-block:: console
+
+   (.venv) $ make epub
```
Comment (ericholscher) [inline doc/tutorial/index.rst:206]: Is there a reason we're doing epub here instead of `pseudoxml`? Just ease of browsing?

File: doc/tutorial/index.rst [hunk 3]
```diff
@@ -184,6 +195,243 @@ as before, or leverage the convenience script as follows:
...
+   so PDF generation is out of scope for this tutorial.
+
+More Sphinx customization using extensions
+------------------------------------------
+
+Enabling a built-in extension
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In addition to these configuration values, you can customize Sphinx even more
+by using :doc:`extensions <../usage/extensions/index>`.  Sphinx ships several
```
Comment (ericholscher) [inline doc/tutorial/index.rst]: Should we use absolute links for these, so that if this content moves we don't have to update them. (Thinking we might break the tutorial up into multiple pages in the future, though I guess they'll still be in the `tutorial/` dir)

File: doc/tutorial/index.rst [hunk 4]
```diff
@@ -184,6 +195,243 @@ as before, or leverage the convenience script as follows:
...
+durations report at the end of the console output, like this one:
+
+.. code-block:: console
+
+   (.venv) $ make html
+   ...
+   The HTML pages are in build/html.
+
+   ====================== slowest reading durations =======================
+   0.042 temp/source/index
```
Comment (ericholscher) [inline doc/tutorial/index.rst:280]: Great example 👍

File: doc/tutorial/index.rst [hunk 5]
```diff
@@ -184,6 +195,243 @@ as before, or leverage the convenience script as follows:
...
+
+Using a third-party theme
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Themes, on the other hand, are a particular class of extensions that allow you
+to customize the appearance of your documentation.  Again, Sphinx has several
```
Comment (ericholscher) [inline doc/tutorial/index.rst]:
```suggestion
to customize the appearance of your documentation.  Sphinx has several
```

File: doc/tutorial/index.rst [hunk 6]
```diff
@@ -184,6 +195,243 @@ as before, or leverage the convenience script as follows:
...
+:ref:`built-in themes <builtin-themes>`, and there are also third-party ones.
```
Comment (ericholscher) [inline doc/tutorial/index.rst]: Link to https://sphinx-themes.org/ for third-party?

File: doc/tutorial/index.rst [hunk 7]
```diff
@@ -184,6 +195,243 @@ as before, or leverage the convenience script as follows:
...
+
+For example, to use the `Furo <https://pradyunsg.me/furo/>`_ third-party theme
+in your HTML documentation, first you will need to install it with ``pip`` on
```
Comment (ericholscher) [inline doc/tutorial/index.rst]:
```suggestion
in your HTML documentation, first you will need to install it with ``pip`` in
```

Comment (ericholscher) [reply doc/tutorial/index.rst]: Seems we're also noting that themes only apply to HTML docs here implicitly, I wonder if this should be more explicit?

File: doc/tutorial/index.rst [hunk 8]
```diff
@@ -184,6 +195,243 @@ as before, or leverage the convenience script as follows:
...
+
+   HTML documentation of Lumache with the Furo theme
+
+Narrative documentation in Sphinx
+---------------------------------
+
+Inserting documents in the project hierarchy
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+As discussed at the beginning, ``index.rst`` is the :term:`master document`,
```
Comment (ericholscher) [inline doc/tutorial/index.rst]: I don't love references to earlier content like this (maybe I don't remember it being discussed? Maybe someone linked me directly to this section for some reason?) -- but don't feel too strongly.

Comment (ericholscher) [reply doc/tutorial/index.rst]: (I also wonder about renaming `master` to `primary` or `index`, but that's another discussion :)

File: doc/tutorial/index.rst [hunk 9]
```diff
@@ -184,6 +195,243 @@ as before, or leverage the convenience script as follows:
...
+   :width: 80%
+   :align: center
+   :alt: HTML documentation of Lumache with the Furo theme
+
+   HTML documentation of Lumache with the Furo theme
+
+Narrative documentation in Sphinx
+---------------------------------
+
+Inserting documents in the project hierarchy
```
Comment (ericholscher) [inline doc/tutorial/index.rst]: I wonder about a more approachable term here. I feel like `hierarchy` isn't really a known concept in terms of documents. Perhaps something like "Structuring your documentation across multiple pages", "Building a menu in the sidebar" or something more goal-oriented.

File: doc/tutorial/index.rst [hunk 10]
```diff
@@ -184,6 +195,243 @@ as before, or leverage the convenience script as follows:
...
+whose main function is to serve as a welcome page and to contain the root of
+the "table of contents tree" (or *toctree*).  Sphinx allows you to assemble
+a project from different files, which is helpful when the project grows.
+
+As an example, create a new file ``docs/source/usage.rst`` (next to
+``index.rst``) with these contents:
+
+.. code-block:: rest
+
+   Usage
```
Comment (ericholscher) [inline doc/tutorial/index.rst:336]:
```suggestion
   # file: docs/source/usage.rst
   
   Usage
```

I find this useful to be doubly explicit. I've seen some themes have a way to actually pull this out into a "header" on the codeblock, which could be a nice benefit in the future. 

Would want to be consistent with this across all the file examples, but might be a good idea.

File: doc/tutorial/index.rst [hunk 11]
```diff
@@ -184,6 +195,243 @@ as before, or leverage the convenience script as follows:
...
+
+This new file contains two :ref:`section <rst-sections>` headers, normal
+paragraph text, and a ``code-block`` directive that renders a block of content
+as source code, with appropriate syntax highlighting (in this case, generic
+``console`` text).
+
+.. note::
+
+   You can read `the list of available highlight
+   languages <https://pygments.org/docs/lexers/>`_ in the Pygments
```
Comment (ericholscher) [inline doc/tutorial/index.rst]: This is super overwhelming.. I wonder if it's not worth including this in the tutorial, but adding it to the reference docs somewhere?

File: doc/tutorial/index.rst [hunk 12]
```diff
@@ -184,6 +195,243 @@ as before, or leverage the convenience script as follows:
...
+   documentation.
+
+The structure of the document is determined by the succession of heading
+styles, which means that, by using ``---`` for the "Installation" section
+after ``===`` for the "Usage" section, you have declared "Installation" to
+be a *subsection* of "Usage".
+
+In addition, add a ``toctree`` :ref:`directive <rst-directives>` at the end
```
Comment (ericholscher) [inline doc/tutorial/index.rst]:
```suggestion
To complete the process, you can add a ``toctree`` :ref:`directive <rst-directives>` at the end
```

A bit more clear that this is the "second step" of the toctree process.

File: doc/tutorial/index.rst [hunk 13]
```diff
@@ -184,6 +195,243 @@ as before, or leverage the convenience script as follows:
...
+
+   .. toctree::
+
+      usage
+
+This step inserts that document in the root of the *toctree*, so now it belongs
+to the structure of your project.  If you export the documentation to HTML
+running ``make html``, you will see that the ``toctree`` gets rendered as a
+list of hyperlinks, and this allows you to navigate to the new page you
+just created.  Neat!
```
Comment (ericholscher) [inline doc/tutorial/index.rst]: I wonder if having a tree diagram for the TOC is useful here -- to help visually show the structure that was created. I find people always get confused with this:

```
index
 - usage
```

File: doc/tutorial/index.rst [hunk 14]
```diff
@@ -184,6 +195,243 @@ as before, or leverage the convenience script as follows:
...
+
+.. warning::
+
+   Every document should belong to a *toctree*.  Otherwise, Sphinx will emit a
+   ``WARNING: document isn't included in any toctree``, and the end result
+   will depend on the builder.  For the HTML builder, the page will not be
+   linked from anywhere (therefore it will not be discoverable), whereas for
+   the PDF builder, it will not be included at all.
```
Comment (ericholscher) [inline doc/tutorial/index.rst]: This warning feels a bit esoteric and complex. I'd remove the PDF section at least, or a generic warning "documents outside the toctree will be impossible for many users to find" or something. 

I'd love to see an "error messages" guide in the docs that have this kind of very explicit content though. Would be great for new users who google errors.

File: doc/tutorial/index.rst [hunk 15]
```diff
@@ -184,6 +195,243 @@ as before, or leverage the convenience script as follows:
...
+
+Adding cross-references
+~~~~~~~~~~~~~~~~~~~~~~~
+
+One powerful feature of Sphinx is the ability to seamlessly add
+:ref:`cross-references <xref-syntax>` to specific parts of the documentation:
+a document, a section, a figure, a code object, etc.  This tutorial is full of
+them!
+
+To add a straightforward cross-reference, write this sentence right after the
```
Comment (ericholscher) [inline doc/tutorial/index.rst]:
```suggestion
To add a cross-reference, write this sentence right after the
```

File: doc/tutorial/index.rst [hunk 16]
```diff
@@ -184,6 +195,243 @@ as before, or leverage the convenience script as follows:
...
+
+Alternatively, you can also add a cross-reference to an arbitrary part of the
+project. For that, you need to use the :rst:role:`ref` role, and add an
+explicit *label* that can act as a target.
+
+For example, to reference the "Installation" subsection, add a label right
+before the heading, as follows:
+
+.. code-block:: rest
+
```
Comment (ericholscher) [inline doc/tutorial/index.rst:420]:
```suggestion
     :emphasize-lines: 4

```

This might be useful.

**Review by ericholscher — COMMENTED** [review]
This looks great! Sorry for all the nitpicks, but mostly just trying to tweak small things.

File: doc/tutorial/index.rst (hunk 10 above)
Comment (ericholscher) [reply doc/tutorial/index.rst:336]: Actually, a `caption` might be the right solution here: https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-option-code-block-caption

File: doc/tutorial/index.rst (hunk 2 above)
Comment (astrojuanlu) [reply doc/tutorial/index.rst:206]: Yes, I went back to your original suggestion of using EPUB, because it's a more familiar format. For beginners, it's unclear what the `pseudoxml` builder is for, and in fact:

> This builder is used for debugging the Sphinx/Docutils “Reader to Transform to Writer” pipeline.

https://www.sphinx-doc.org/en/master/usage/builders/index.html#sphinx.builders.xml.PseudoXMLBuilder

The good thing is that, even if the user doesn't have an EPUB viewer, they can still open the index in a web browser.

File: doc/tutorial/index.rst (hunk 16 above)
Comment (astrojuanlu) [reply doc/tutorial/index.rst:420]: Love this! Will use it more extensively

File: doc/tutorial/index.rst (hunk 3 above)
Comment (astrojuanlu) [reply doc/tutorial/index.rst]: Good idea, changing it

File: doc/tutorial/index.rst (hunk 6 above)
Comment (astrojuanlu) [reply doc/tutorial/index.rst]: Good idea 💪🏽

File: doc/tutorial/index.rst (hunk 8 above)
Comment (astrojuanlu) [reply doc/tutorial/index.rst]: You're right. (As I read the other day: "Never assume that anyone would read your article word to word, from beginning to end".) Will rephrase it.

About "master document", yeah, that's for another day :)

File: doc/tutorial/index.rst (hunk 11 above)
Comment (astrojuanlu) [reply doc/tutorial/index.rst]: True, I added this to the `code-block` documentation instead 👍🏽

File: doc/tutorial/index.rst (hunk 14 above)
Comment (astrojuanlu) [reply doc/tutorial/index.rst]: True, it's too long. Will rephrase it to make it simpler.

Comment (astrojuanlu) [reply doc/tutorial/index.rst]: About the "errors" section, I think the main issue is that `make ...` produces lots of colored output, and warnings that should be addressed pass silently. It's a difficult problem I think, but the case of out-of-doctree documents should be treated as an error in my opinion (with an option to ignore it).

Comment (astrojuanlu) [reply doc/tutorial/index.rst]: (I'm 👍🏽 on adding a FAQ some day)

File: doc/tutorial/index.rst (hunk 7 above)
Comment (astrojuanlu) [reply doc/tutorial/index.rst]: I wonder if we should address that in a separate PR, since all the documentation seems to be suggesting that. Every "theme" reference in https://www.sphinx-doc.org/en/master/contents.html is for "HTML" ("HTML Theming", "HTML theme development", "Documentation using a custom theme or integrated in a website"). While writing this response, I tried searching for "sphinx pdf theme", and I got [a Stack Overflow answer saying "this can't be done"](https://stackoverflow.com/a/9806839/554319), an [unhelpful thread in sphinx-users](https://groups.google.com/g/sphinx-users/c/LBGxmGESUhg), and finally a [quite new PDF theme](https://pypi.org/project/sphinx-business-theme/) that however doesn't even use LaTeX, but [weasyprint](https://pypi.org/project/weasyprint/) (¡hola @nekmo! 👋🏽)

Even if it's possible to have Sphinx themes for things other than HTML, I would like to not point people to things they can't easily find at the moment, without rethinking other parts of the Sphinx documentation first.

Comment (astrojuanlu) [reply doc/tutorial/index.rst]: In any case, I changed the header to say "HTML theme" to make room for other kind of themes in the future 👍🏽

File: doc/tutorial/index.rst (hunk 13 above)
Comment (astrojuanlu) [reply doc/tutorial/index.rst]: Added a small text representation of the structure.

File: doc/tutorial/index.rst (hunk 10 above)
Comment (astrojuanlu) [reply doc/tutorial/index.rst:336]: Nice! Adding this to all files

File: doc/tutorial/index.rst (hunk 9 above)
Comment (astrojuanlu) [reply doc/tutorial/index.rst]: +100 for "something more goal-oriented" 👍🏽 your first suggestion looks good to me. Renamed the header.

**Comment by astrojuanlu** [comment]
Thank you @ericholscher for the thorough review! I have addressed your comments.

File: doc/tutorial/index.rst [hunk 17]
```diff
@@ -184,6 +197,250 @@ as before, or leverage the convenience script as follows:
...
+Sphinx supports a variety of formats apart from HTML, including PDF, EPUB,
+:ref:`and more <builders>`.  For example, to build your documentation as an
+e-book in EPUB format, run this command from the ``docs`` directory:
+
+.. code-block:: console
+
+   (.venv) $ make epub
+
+After that, you will see the files corresponding to the e-book under
+``docs/build/html/``.  You can either open ``Lumache.epub`` with an
```
Comment (humitos) [inline doc/tutorial/index.rst]:
```suggestion
``docs/build/epub/``.  You can either open ``Lumache.epub`` with an
```

I think the directory is build/epub.

**Review by humitos — APPROVED** [review]
Good progress on this!

File: doc/tutorial/index.rst (hunk 17 above)
Comment (astrojuanlu) [reply doc/tutorial/index.rst]: Oops, good catch! Thanks @humitos 🙏🏽

File: doc/tutorial/index.rst [hunk 18]
```diff
@@ -184,6 +197,250 @@ as before, or leverage the convenience script as follows:
...
+.. code-block:: python
+
+   # EPUB options
+   epub_show_urls = 'footnote'
+
+With this configuration value, and after running ``make epub`` again, you will
+notice that URLs appear now as footnotes, which avoids cluttering the text.
+Sweet!
+
+.. warning::
```
Comment (jakobandersen) [inline doc/tutorial/index.rst]:
```suggestion
.. note::
```
A warning seems a bit too strong to me.

File: doc/tutorial/index.rst [hunk 19]
```diff
@@ -184,6 +197,250 @@ as before, or leverage the convenience script as follows:
...
+
+   Generating a PDF using Sphinx can be done running ``make latexpdf``,
+   provided that the system has a working :math:`\LaTeX` installation,
+   as explained in the documentation of :class:`sphinx.builders.latex.LaTeXBuilder`.
+   Although this is perfectly feasible, such installations are often big,
+   and in general LaTeX requires careful configuration in some cases,
+   so PDF generation is out of scope for this tutorial.
+
+More Sphinx customization using extensions
```
Comment (jakobandersen) [inline doc/tutorial/index.rst]:
```suggestion
More Sphinx customization
```
See comment below on HTML themes.

File: doc/tutorial/index.rst [hunk 20]
```diff
@@ -184,6 +197,250 @@ as before, or leverage the convenience script as follows:
...
+------------------------------------------
+
```
Comment (jakobandersen) [inline doc/tutorial/index.rst:247]: I suggest writing a few sentences here to introduce the different kinds of customization.

File: doc/tutorial/index.rst [hunk 21]
```diff
@@ -184,6 +197,250 @@ as before, or leverage the convenience script as follows:
...
+   ...
+   The HTML pages are in build/html.
+
+   ====================== slowest reading durations =======================
+   0.042 temp/source/index
+
+Using a third-party HTML theme
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Themes, on the other hand, are a particular class of extensions that allow you
```
Comment (jakobandersen) [inline doc/tutorial/index.rst]: While a theme is an extension in the abstract sense, I think the term "extension" for Sphinx is mostly used in a technical sense about a Python module which must be added to the ``extensions`` conf var.
(ping @tk0miya for potential corrections of understanding)

File: doc/tutorial/index.rst [hunk 22]
```diff
@@ -184,6 +197,250 @@ as before, or leverage the convenience script as follows:
...
+   HTML documentation of Lumache with the Furo theme
+
+Narrative documentation in Sphinx
+---------------------------------
+
+Structuring your documentation across multiple pages
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The file ``index.rst`` created by ``sphinx-quickstart`` is the :term:`master
+document`, whose main function is to serve as a welcome page and to contain the
```
Comment (jakobandersen) [inline doc/tutorial/index.rst:325]: In 4.0 "master" was changed to "root", https://www.sphinx-doc.org/en/master/usage/configuration.html#confval-root_doc

File: doc/tutorial/index.rst [hunk 23]
```diff
@@ -184,6 +197,250 @@ as before, or leverage the convenience script as follows:
...
+   :caption: docs/source/index.rst
+
+   Check out the :doc:`usage` section for further information, including how to
+   :ref:`install <installation>` the project.
+
+Notice a trick here: the ``install`` part specifies how the link will look like
+(we want it to be a specific word, so the sentence makes sense), whereas the
+``<installation>`` part refers to the actual label we want to add a
+cross-reference to. Both the ``:doc:`` and the ``:ref:`` roles will be rendered
+as hyperlinks in the HTML documentation.
```
Comment (jakobandersen) [inline doc/tutorial/index.rst:445]: Is this the first time the ``:ref:`` role is explained? If so I think it would be good to explain that the title is optional, and what happens when none is given.

**Review by jakobandersen — CHANGES_REQUESTED** [review]

File: doc/tutorial/index.rst (hunk 18 above)
Comment (astrojuanlu) [reply doc/tutorial/index.rst]: You're right, accepting

File: doc/tutorial/index.rst (hunk 22 above)
Comment (astrojuanlu) [reply doc/tutorial/index.rst:325]: Ah, great to know! Changing in the rest of the document as well

File: doc/tutorial/index.rst (hunk 21 above)
Comment (astrojuanlu) [reply doc/tutorial/index.rst]: According to the Glossary, "A custom role, directive or other aspect of Sphinx that allows users to modify any aspect of the build process within Sphinx." https://www.sphinx-doc.org/en/master/glossary.html#term-extension

And then, according to https://www.sphinx-doc.org/en/master/usage/extensions/index.html, "Sphinx allows adding “extensions” to the build process, each of which can modify almost any aspect of document processing.".

In any case, I get your point: that themes are not extensions, but another form of customization. Will rephrase this.

File: doc/tutorial/index.rst (hunk 20 above)
Comment (astrojuanlu) [reply doc/tutorial/index.rst:247]: Done! (Added a brief introductory sentence)

File: doc/tutorial/index.rst (hunk 19 above)
Comment (astrojuanlu) [reply doc/tutorial/index.rst]: Done locally

File: doc/tutorial/index.rst (hunk 23 above)
Comment (astrojuanlu) [reply doc/tutorial/index.rst:445]: Good idea! Changed

**Review by YoutacVA — APPROVED** [review]

**Comment by astrojuanlu** [comment]
@jakobandersen @tk0miya @jfbu Hi folks, I wonder if you have any extra thoughts on this as well as #9341. In the coming days I will start working on part 3 of the tutorial, so I would love to know if you think there are any blockers to get going. If it's just a matter of finding the time and will to review and merge the pull request, I understand 🙏🏽

**Comment by tk0miya** [comment]
Sorry, I did not have time to review this in recent weeks. I'll take time for this in this week. Please wait a moment. I think there is nothing to block your next work. Thank you for your great work!

File: doc/glossary.rst [hunk 24]
```diff
@@ -73,7 +73,7 @@ Glossary
 
      For more information, refer to :doc:`/usage/extensions/index`.
 
-   master document
```
Comment (tk0miya) [inline doc/glossary.rst:76]: My recommendation is keeping this as a term referencing to the "root document".

I think it's okay to use the "root document" instead of "master document". It's a better name for me because we've already used the "tree" metaphor for ToC, and "root" is much familiar with the model. But we need to keep the old name because it has been used very very longly. I guess the terminology would be used in many blog articles. And, it might be referenced via intersphinx.

File: doc/tutorial/index.rst [hunk 25]
```diff
@@ -27,6 +27,9 @@ a basic understanding of how it works, as well as a working Python installation
 for development, since you will use *Python virtual environments* to create the
 project.
 
+.. contents:: Contents
```
Comment (tk0miya) [inline doc/tutorial/index.rst]: Is this needed? We already have local ToC table on right side bar. No reason to show it, I think.

File: doc/tutorial/index.rst [hunk 26]
```diff
@@ -184,6 +197,254 @@ as before, or leverage the convenience script as follows:
 After running this command, you will see that ``index.html`` reflects the new
 changes!
 
+Exporting your documentation to other formats
```
Comment (tk0miya) [inline doc/tutorial/index.rst]: Is it better to use "export" instead of "build"? AFAIK, Sphinx is usually described as a "documentation builder". So it would be better to use "build". Personally, I still don't understand what is better wording to this behavior; build, convert or others (Because I'm not a native speaker!).

File: doc/usage/extensions/index.rst [hunk 27]
```diff
@@ -10,6 +10,8 @@ This chapter describes the extensions bundled with Sphinx.  For the API
 documentation on writing your own extension, refer to :ref:`dev-extensions`.
 
 
+.. _built-in-extensions:
```
Comment (tk0miya) [inline doc/usage/extensions/index.rst]: Which is better prefix; `builtin-` or `built-in-`? It seems `builtin-themes` has been already defined. So this should be renamed to `builtin-`, or renaming it to `built-in-` also.

File: doc/usage/restructuredtext/directives.rst [hunk 28]
```diff
@@ -488,6 +488,8 @@ __ https://pygments.org/docs/lexers
 
       .. versionadded:: 2.1
 
+.. _rst-code-block:
```
Comment (tk0miya) [inline doc/usage/restructuredtext/directives.rst]: No labels are needed here. You can refer the `code-block` directive via `:rst:dir:` role.

File: doc/tutorial/index.rst [hunk 29]
```diff
@@ -184,6 +197,254 @@ as before, or leverage the convenience script as follows:
...
+
+This step inserts that document in the root of the *toctree*, so now it belongs
+to the structure of your project, which so far looks like this:
+
+.. code-block:: text
+
+   index
+   └── usage
+
+If you export the documentation to HTML running ``make html``, you will see
```
Comment (tk0miya) [inline doc/tutorial/index.rst]: export?

File: doc/tutorial/index.rst [hunk 30]
```diff
@@ -184,6 +197,254 @@ as before, or leverage the convenience script as follows:
...
+   :caption: docs/source/index.rst
+
+   Check out the :doc:`usage` section for further information.
+
+The :rst:role:`doc` role you used automatically references a specific document
+in the project, in this case the ``usage.rst`` you created earlier.
+
+Alternatively, you can also add a cross-reference to an arbitrary part of the
+project. For that, you need to use the :rst:role:`ref` role, and add an
+explicit *label* that can act as a target.
```
Comment (tk0miya) [inline doc/tutorial/index.rst]: FYI: We've called this notation "label" longly. But it has been called "Hyperlink targets" notation formally. So it's natural to act as a target :-)
https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#hyperlink-targets

File: doc/tutorial/index.rst [hunk 31]
```diff
@@ -184,6 +197,254 @@ as before, or leverage the convenience script as follows:
...
+
+For example, to reference the "Installation" subsection, add a label right
+before the heading, as follows:
+
+.. code-block:: rest
```
Comment (tk0miya) [inline doc/tutorial/index.rst]: We usually uses `rst` as a highlight language for reStructuredText.

```
tkomiya@Alrescha> grep -r code-block:: doc | grep ':: rst'
doc/man/sphinx-autogen.rst:.. code-block:: rst
doc/development/theming.rst:.. code-block:: rst
doc/development/tutorials/todo.rst:.. code-block:: rst
doc/development/tutorials/todo.rst:.. code-block:: rst
doc/development/tutorials/todo.rst:.. code-block:: rst
doc/development/tutorials/helloworld.rst:.. code-block:: rst
doc/development/tutorials/helloworld.rst:.. code-block:: rst
doc/development/tutorials/recipe.rst:.. code-block:: rst
doc/development/tutorials/recipe.rst:.. code-block:: rst
doc/development/tutorials/autodoc_ext.rst:.. code-block:: rst
doc/usage/configuration.rst:   .. code-block:: rst
doc/usage/extensions/doctest.rst:.. code-block:: rst
doc/faq.rst:  .. code-block:: rst
```

I noticed some code-blocks uses "rest"...
```
tkomiya@Alrescha> grep -r code-block:: doc | grep ':: rest'
doc/tutorial/index.rst:.. code-block:: rest
doc/tutorial/index.rst:.. code-block:: rest
doc/usage/quickstart.rst:.. code-block:: rest
doc/usage/quickstart.rst:.. code-block:: rest
doc/usage/quickstart.rst:.. code-block:: restructuredtext
doc/usage/quickstart.rst:.. code-block:: restructuredtext
```

Comment (tk0miya) [reply doc/tutorial/index.rst]: Note: Both `rest` and `rst` are supported.
https://pygments.org/docs/lexers/#pygments.lexers.markup.RstLexer

File: doc/tutorial/index.rst (hunk 26 above)
Comment (jakobandersen) [reply doc/tutorial/index.rst]: Good point. I think it is worthwhile being consistent with the existing interface and use the term "building" (or whichever variantion fits the context): a new format is added by creating a [Builder](https://www.sphinx-doc.org/en/master/usage/builders/index.html) extension and you select the output format with ``-b`` on the command line.
As another alternative to "export" I suggest "compile". To me "export" seems more like a direct write in a different format, but Sphinx really is performing tasks akin to an ordinary compiler.

File: doc/tutorial/index.rst [hunk 32]
```diff
@@ -134,23 +138,32 @@ the documentation as HTML for the first time.  To do that, run this command:
...
+   Freshly created documentation of Lumache
 
 There we go! You created your first HTML documentation using Sphinx.
 
-Making some tweaks to the index
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+First steps to document your project using Sphinx
+-------------------------------------------------
+
+Converting your documentation to HTML
```
Comment (jakobandersen) [inline doc/tutorial/index.rst]: See next comment about "export" vs. "build" vs. "compile". "Converting" should probably change accordingly.

File: doc/glossary.rst (hunk 24 above)
Comment (astrojuanlu) [reply doc/glossary.rst:76]: Fair enough, I will rephrase this so we can include both and slowly phase out the old name.

Comment (astrojuanlu) [reply doc/glossary.rst:76]: Oh sorry, this is in the glossary. Understood.

File: doc/tutorial/index.rst (hunk 25 above)
Comment (astrojuanlu) [reply doc/tutorial/index.rst]: Fair enough. I added it because it looked cool and I didn't know it was possible before I started writing :)

File: doc/tutorial/index.rst (hunk 26 above)
Comment (astrojuanlu) [reply doc/tutorial/index.rst]: Good points, I will rephrase to "build". "Compile" is too attached to machine instructions, I think.

Comment (astrojuanlu) [reply doc/tutorial/index.rst]: I was tempted to introduce a preposition that's probably incorrect ("build your docs AS html", "build your docs TO html"), so instead I kept "build your HTML docs".

Comment (astrojuanlu) [reply doc/tutorial/index.rst]: (Forget my last comment - sometimes I think it makes sense to say "build docs AS x")

File: doc/tutorial/index.rst (hunk 29 above)
Comment (astrojuanlu) [reply doc/tutorial/index.rst]: ✔️

File: doc/usage/restructuredtext/directives.rst (hunk 28 above)
Comment (astrojuanlu) [reply doc/usage/restructuredtext/directives.rst]: Right, thanks!

File: doc/tutorial/index.rst (hunk 32 above)
Comment (astrojuanlu) [reply doc/tutorial/index.rst]: Addressed, thanks!

File: doc/tutorial/index.rst (hunk 31 above)
Comment (astrojuanlu) [reply doc/tutorial/index.rst]: Fair enough, I will convert everything to `rst` for consistency.

File: doc/usage/extensions/index.rst (hunk 27 above)
Comment (astrojuanlu) [reply doc/usage/extensions/index.rst]: According to the Merrian-Webster dictionary, "builtin" doesn't exist and it should be "built-in". So, we got it wrong in other places :) Will change it.

Comment (astrojuanlu) [reply doc/usage/extensions/index.rst]: On second thought, we are using "builtin" in too many places. I will rename to `builtin-` for consistency, we can decide in another PR.

File: doc/tutorial/index.rst (hunk 30 above)
Comment (astrojuanlu) [reply doc/tutorial/index.rst]: Thanks for the clarification :) Will use this opportunity to link to the reST docs.

File: doc/tutorial/index.rst [hunk 33]
```diff
@@ -184,6 +194,256 @@ as before, or leverage the convenience script as follows:
...
+   Check out the :doc:`usage` section for further information.
+
+The :rst:role:`doc` role you used automatically references a specific document
+in the project, in this case the ``usage.rst`` you created earlier.
+
+Alternatively, you can also add a cross-reference to an arbitrary part of the
+project. For that, you need to use the :rst:role:`ref` role, and add an
+explicit *label* that acts as `a target`__.
+
+__ https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#hyperlink-targets
```
Comment (astrojuanlu) [inline doc/tutorial/index.rst:412]: This causes `tox -e docslint` to fail. I don't know of any way to prevent that.

**Comment by astrojuanlu** [comment]
Thanks folks for the extra round of reviews! I addressed all the comments. `docslint` is failing because there is a long URL that I don't know how to split, other than that this is ready I think.

File: doc/tutorial/index.rst (hunk 33 above)
Comment (tk0miya) [reply doc/tutorial/index.rst:412]: Oops. Additional rule is needed for this case: https://github.com/sphinx-doc/sphinx/pull/9405

**Review by tk0miya — COMMENTED** [review]
LGTM!

**Review by tk0miya — APPROVED** [review]

File: doc/tutorial/index.rst (hunk 33 above)
Comment (astrojuanlu) [reply doc/tutorial/index.rst:412]: Merged `4.x` on this branch to include #9405, thanks!

**Review by jakobandersen — APPROVED** [review]
LGTM

**Comment by tk0miya** [comment]
Just merged. Thank you for your great work!

**Comment by astrojuanlu** [comment]
Thanks a lot @tk0miya! On to the next PR 💪🏽

**Comment by ericholscher** [comment]
Excited to have this moving forward. Great work everyone. <3

File: doc/tutorial/index.rst [hunk 34]
```diff
@@ -184,6 +194,256 @@ as before, or leverage the convenience script as follows:
...
+   epub_show_urls = 'footnote'
+
+With this configuration value, and after running ``make epub`` again, you will
+notice that URLs appear now as footnotes, which avoids cluttering the text.
+Sweet!
+
+.. note::
+
+   Generating a PDF using Sphinx can be done running ``make latexpdf``,
+   provided that the system has a working :math:`\LaTeX` installation,
```
Comment (jfbu) [inline doc/tutorial/index.rst:236]: This breaks PDF production, as `\(\LaTeX\)` mark-up is invalid. `\LaTeX` can not be used in math mode.

Comment (astrojuanlu) [reply doc/tutorial/index.rst:236]: Woops, sorry. Let's continue the conversation in #9475 and #9476
